music_rnn/src/model.ts

```diff
@@ -131,41 +155,160 @@ export class MusicRNN {
 
     const oh = tf.tidy(() => {
       const inputs = this.dataConverter.toTensor(sequence);
-
-      const length:number = inputs.shape[0];
       const outputSize:number = inputs.shape[1];
-
-      let c: tf.Tensor2D[] = [];
-      let h: tf.Tensor2D[] = [];
-      for (let i = 0; i < this.biasShapes.length; i++) {
-        c.push(tf.zeros([1, this.biasShapes[i] / 4]));
-        h.push(tf.zeros([1, this.biasShapes[i] / 4]));
-      }
-
-      // Initialize with input.
-      const samples: tf.Tensor1D[] = [];
-      for (let i = 0; i < length + steps; i++) {
-        let nextInput: tf.Tensor2D;
-        if (i < length) {
-          nextInput = inputs.slice([i, 0], [1, outputSize]).as2D(1, outputSize);
-        } else {
-          const logits = h[h.length - 1].matMul(this.lstmFcW).add(this.lstmFcB);
-          const sampledOutput = (
-            temperature ?
-            tf.multinomial(logits.div(tf.scalar(temperature)), 1).as1D():
-            logits.argMax(1).as1D());
-          nextInput = tf.oneHot(sampledOutput, outputSize).toFloat();
-          // Save samples as bool to reduce data sync time.
-          samples.push(nextInput.as1D().toBool());
-        }
-        [c, h] = tf.multiRNNCell(this.lstmCells, nextInput, c, h);
-      }
-
+      const samples = this.attnMatrix ?
+        this.sampleRnnWithAttention(inputs, steps, temperature) :
+        this.sampleRnn(inputs, steps, temperature);
       return tf.stack(samples).as2D(samples.length, outputSize);
```

Comment:
These implementations are different enough that might be appropriate to split them out into separate classes. In Python there's a standalone `AttentionCellWrapper` that manages attention transparently. Using such a thing here may require changing the abstractions a bit.

Reply:
Agreed. Let's split this into two PRs. Just make the drum converter change in this PR for now.